Add thumbnail optimiser for avatar upload

willm30 opened the pull request:
## Description

This PR adds support for thumbnail optimisation when a user uploads an avatar to their profile in User Settings. 

![avatar upload](https://user-images.githubusercontent.com/64402732/210350647-c9a71a6a-05b7-4e86-9392-beff8974b2e5.gif)

It also refactors the AvatarUploader to use a new component `SingleFileUpload`, instead of the original `FileUpload` (see heading below for explanation). 

## Test

Upload an avatar. The thumbnail should automatically update per GIF above. None of the `AvatarUpload` functionality should have changed so please also double check this, e.g. upload an image that's too large / wrong file type, remove the avatar etc.  

**New stuff** ✨

* `getThumbnail` (thumbnail optimiser)
* `SingleFileUpload`

**Changes** 🏗

* Mainly made changes to the `AvatarUploader` and `UserAvatarUploader`, to accomodate the new components above.
 
**Deletions** ⚰️

* `AvatarUploadItem` -- no longer necessary.
* `Sidebar` -- was a fairly superfluous component and has been merged with `UserAvatarUploader`.

### Thumbnail Optimisation 

I looked into a few different libraries, but in the end went with [Pica](https://github.com/nodeca/pica), which has a bundle size of 216kb and 3.2k stars on github. Specifically, I'm using a [wrapper](https://github.com/nodeca/image-blob-reduce) around Pica also created by the Pica author, to resize images from Blobs/Files (as opposed to image / canvas elements). The image optimisation algorithm used is [`mks2013`](https://johncostella.com/magic/) and I am only using the pica defaults.
 
### SingleFileUpload

`FileUpload` is designed with the CSV uploader in mind, or in other words, designed to support the uploading of multiple files and the dynamic rendering of a component per file uploaded (see `CSVUploaderItem` for an example). 

This is not the same use case as the `AvatarUploader`, which only needs one file uploading and nothing to be dynamically rendered once uploaded. I think this distinction is sufficient to warrant two separate components (not least because it makes it easier and simpler to understand what's happening now when uploading an avatar). 

In order to avoid duplication of the actual uploading logic, I have created the `useDropzoneWithFileReader` hook, which separates the file reading/uploading logic from the file reading view. Not only does it make the upload process easier to follow, it also makes the file uploading indifferent as to whether it's rendered inside of a form or not (which I expect, again, is more useful for uploading multiple files than a single file).

Although I haven't done it in this pr, I envisage `useDropzoneWithFileReader` also being used in a refactored `FileUpload` when the `CSVUploader` is implemented (becoming `MultipleFileUpload` in the process).  

Ultimately, the intention here is to abstract uploading logic so it's separated from the upload view, which allows us to have separate components for single file uploads with no dynamic content versus multiple file uploads with dynamic content. 

Resolves  #141, #142

willm30 commented on src/components/shared/UserAvatar/UserAvatar.tsx:
```diff
@@ -35,7 +35,6 @@ const UserAvatar = ({
   address,
   banned = false,
   colony,
-  notSet = true,
```
Removing this default as more often than not it was being overridden

willm30 commented on src/hooks/useDropzoneWithFileReader.ts:
```diff
@@ -0,0 +1,50 @@
+import { HandleFileAccept } from '~shared/FileUpload/types';
+import getFileReader from '~utils/fileReader';
+
+const useDropzoneWithFileReader = ({
```
This hook makes the use of dropzone more flexible and separates the logic from a specific view.

willm30 commented on src/images/optimisation/index.ts:
```diff
@@ -0,0 +1,31 @@
+import ImageBlobReduce, { Options } from 'image-blob-reduce';
```
`image-blob-reduce` is a wrapper around the [pica library](https://github.com/nodeca/pica). Under the hood, it uses the [`mks2013`](https://johncostella.com/magic/) algorithm to handle image optimisation, and handles resizing via the use of a `canvas` element.

willm30 commented on src/types/definitions/image-blob-reduce/index.d.ts:
```diff
@@ -0,0 +1,33 @@
+declare module 'image-blob-reduce' {
```
Manually added types as `image-blob-reduce` doesn't come with any

willm30 commented on src/utils/fileReader/index.ts:
```diff
@@ -34,22 +39,31 @@ const fileReaderFactory = (options: Partial<FileReaderOptions>) => {
+        let name: File['name'] | undefined;
+        let lastModified: File['lastModified'] | undefined;
+
+        if (file instanceof File) {
```
fileReader can also handle blobs, and in the case of thumbnail optimisation, a `Blob` is returned by the optimiser. Blobs don't have `name` or `lastModified` fields, hence adding this check.

willm30 commented on src/utils/fileReader/index.ts:
```diff
@@ -34,22 +39,31 @@ const fileReaderFactory = (options: Partial<FileReaderOptions>) => {
           lastModified,
           uploadDate: Date.now(),
           data: evt.target.result,
+          file,
```
Having access to the original input file is useful for thumbnail optimisation. (see `UserAvatarUploader: handleFileUpload`)

willm30 commented on src/utils/fileReader/index.ts:
```diff
@@ -21,7 +26,7 @@ const fileReaderFactory = (options: Partial<FileReaderOptions>) => {
   }
 
-  function readFilePromise(file: File): Promise<FileReaderFile | Error> {
+  function readFile(file: File | Blob): Promise<FileReaderFile> {
```
Renaming as after adding types to this document, including the return type in the function name is no longer helpful

willm30 commented on src/components/shared/FileUpload/SingleFileUpload.tsx:
```diff
@@ -0,0 +1,63 @@
+
+const displayName = 'SingleFileUpload';
+
+const SingleFileUpload = ({
```
I believe this is easier to follow than using the `FileUpload` component, which is tailored more for the CSV uploader than the Avatar uploader.

Julianahlk commented on src/components/common/UserProfileEdit/UserAvatarUploader.tsx:
```diff
@@ -1,20 +1,27 @@
   uploaderLabel: {
     id: `${displayName}.uploaderLabel`,
-    defaultMessage: 'At least 250x250px, up to 1MB, .png or .svg',
+    defaultMessage: 'At least 250x250px, up to 280KB, .png or .svg',
```
@willm30 280kb is a bit small and seems like an inconvenience. Couldn't the image be compressed  more so a large initial file size could be used?

jakubcolony replied:
I agree with Juliana this limit is quite restrictive. In longer term we might want to consider storing static assets in S3, there already seems to be a dedicated Amplify library for this - https://docs.amplify.aws/lib/storage/getting-started/q/platform/js/.

Julianahlk commented on src/components/shared/AvatarUploader/AvatarUploader.tsx:
```diff
@@ -34,115 +34,146 @@ const ACCEPTED_MIME_TYPES = {
   'image/png': [],
```
@willm30  Its always been like this but since you are making changes, is there any reason not to allow uploading jpeg?

Julianahlk reviewed (approved):
Had a few comments but overall looks good and working well for me so approving. 💯

willm30 commented:
@Julianahlk I think that's a good suggestion. I've added jpg and webp support, and removed the max file size since we can compress the avatar in the browser before uploading to the db. I've also removed the  label above the avatar since it seems redundant (e.g. is there any reason to say min 250px x 250? This wasn't being enforced in the dapp.)

Julianahlk commented:
> @Julianahlk I think that's a good suggestion. I've added jpg and webp support, and removed the max file size since we can compress the avatar in the browser before uploading to the db. I've also removed the label above the avatar since it seems redundant (e.g. is there any reason to say min 250px x 250? This wasn't being enforced in the dapp.)

@willm30 after your changes, what happens now is if I attempt to upload a png file that is larger than ~1.1MB I am getting the error "Accepted file types are: svg, jpg, png and webp". However, jpeg files of all sizes seem to work fine (I have tried up to 20MB).

willm30 commented:
@Julianahlk Hmm, I think I'll need to test with the image you're using to see exactly what's happening since I've been testing with a PNG of 10mb and having no problem. I've even resized it to 1.1MB just to be sure and am having no trouble. 

Sometimes the compression actually makes the file larger, which may be the case here. 

If you want you could also go to `UserAvatarUploader` and log out the errors in `onError` on line 40 and `handleFileReject` on line 72. Both get passed the error as the first parameter and these are the only two places an error could be coming from.

jakubcolony reviewed (approved):
Some really good quality work @willm30! 

The only thing that stood out is that if you try to upload a corrupted file (e.g. change the extension of a file from pdf to jpg), there's an uncaught ImageBlobReduce error in the console and the spinner just keeps on going:
```
Uncaught (in promise) Error: ImageBlobReduce: failed to create Image() from blob
```
Maybe we could also handle this case and show a nice error message?

willm30 commented:
@jakubcolony @Julianahlk Thanks both for helping find these edge cases! Given that the upload size error doesn't seem to be an issue with removing the max limit, but instead the unpredictable result of compressing a given image, I've opted to improve the error messaging instead of reimposing the old limit of 1mb. 

For images that fail to compress sufficiently:
![toolarge](https://user-images.githubusercontent.com/64402732/211323783-ae7fe85c-15ae-4ef3-acfe-53d542965fe7.png)

For images that are the incorrect file type : 
![unsupported](https://user-images.githubusercontent.com/64402732/211323797-775f4c79-eb66-4c17-9c11-63f19a642e73.png)

And in all other cases: 
![corrupt](https://user-images.githubusercontent.com/64402732/211323807-b74dc6c9-7193-403f-be7c-9dfb93e04f76.png)

jakubcolony reviewed (approved):
Great stuff, everything works now! Or if it doesn't, it still looks nice:
<img width="264" alt="image" src="https://user-images.githubusercontent.com/112586815/211416577-88982980-1e47-4b39-83f5-63cfa50c9743.png">
<img width="256" alt="image" src="https://user-images.githubusercontent.com/112586815/211416636-d70c0cf8-6095-473f-b331-dec1293c34ff.png">

Julianahlk commented:
@willm30 Tested again and getting the same results as Jakub so its all good now. Overall, it seems like jpegs compress much better than png images.
